langchain/chains/tot/base.py

```diff
@@ -0,0 +1,317 @@
+    output_key: str = "response"  #: :meta private:
+    k: int = 10
+    """The maximmum number of conversation rounds"""
+    tot_memory: ToTMemory = ToTMemory()
+    tot_controller: ToTController = ToTController()
+
+    class Config:
+        """Configuration for this pydantic object."""
+
+        extra = Extra.forbid
+        arbitrary_types_allowed = True
+
+    @property
+    def tot_prompter(self) -> ToTPrompter:
+        return ToTPrompter(
+            first_step_prompt=self.first_step_prompt,
+            next_step_prompt=self.next_step_prompt,
+        )
+
+    @property
+    def input_keys(self) -> List[str]:
+        """Will be whatever keys the prompt expects.
+
+        :meta private:
+        """
+        return self.first_step_prompt.input_variables
+
+    @property
+    def output_keys(self) -> List[str]:
+        """Will always return text key.
+
+        :meta private:
+        """
+        return [self.output_key]
+
+    def parse_llm_output(self, llm_output: str) -> str:
+        """Parse the output of the language model."""
+        if match := re.search(r"\{.*?\}", llm_output, re.DOTALL):
+            try:
+                return json.loads(match.group())["next_step"]
+            except json.JSONDecodeError:
+                return ""
+        return ""
+
+    @property
+    def current_level(self) -> int:
+        return len(self.tot_memory.stack)
+
+    def log_request(
+        self, prompt: str, run_manager: Optional[CallbackManagerForChainRun] = None
+    ) -> None:
+        if run_manager is not None:
+            prefix = "    " * self.current_level
+            run_manager.on_text(
+                indent(f"ToT Request >>>\n{prompt}\n\n", prefix),
+                verbose=self.verbose,
+                color="green",
+            )
+
+    def log_response(
+        self, response: str, run_manager: Optional[CallbackManagerForChainRun] = None
+    ) -> None:
+        if run_manager is not None:
+            prefix = "    " * self.current_level
+            run_manager.on_text(
+                indent(f"ToT Response >>>\n{response}\n\n", prefix),
+                verbose=self.verbose,
+                color="green",
+            )
+
+    def log_ctrl_signal(
+        self,
+        ctrl_signal: Optional[str],
+        run_manager: Optional[CallbackManagerForChainRun] = None,
+    ) -> None:
+        if ctrl_signal is None:
+            ctrl_signal = "<None>"
+        if run_manager is not None:
+            prefix = "    " * self.current_level
+            run_manager.on_text(
+                indent(f"ToT Ctrl Signal >>>\n{ctrl_signal}\n\n", prefix),
+                verbose=self.verbose,
+                color="red",
+            )
+
+    def solve(
+        self,
+        problem_description: str,
+        run_manager: Optional[CallbackManagerForChainRun] = None,
+    ) -> Optional[str]:
+        """Algorithm 2 from the ToT paper."""
+        checker = partial(self.checker, problem_description)
```

Comment:
nit: not sure these lines add much

Reply:
I was trying to write something that looked very close to the algorithm in the paper. But I agree that it doesn't make it more readable.